src/cast.c

```diff
@@ -0,0 +1,271 @@
+    double elt = *data;
+
+    if (elt != 0 && elt != 1) {
+      *lossy = true;
+      UNPROTECT(1);
+      return R_NilValue;
+    }
+
+    *out_data = isnan(elt) ? NA_LOGICAL : (int) elt;
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP chr_as_logical(SEXP x, bool* lossy) {
+  SEXP* data = STRING_PTR(x);
+  R_len_t n = Rf_length(x);
+
+  SEXP out = PROTECT(Rf_allocVector(LGLSXP, n));
+  int* out_data = LOGICAL(out);
+
+  for (R_len_t i = 0; i < n; ++i, ++data, ++out_data) {
+    const char* elt = CHAR(*data);
+    switch (elt[0]) {
+    case 'T':
+      if (elt[1] == '\0' || strcmp(elt, "TRUE") == 0) {
+        *out_data = 1;
+        continue;
+      }
+      break;
+    case 'F':
+      if (elt[1] == '\0' || strcmp(elt, "FALSE") == 0) {
+        *out_data = 0;
+        continue;
+      }
+      break;
+    case 't':
+      if (strcmp(elt, "true") == 0) {
+        *out_data = 1;
+        continue;
+      }
+      break;
+    case 'f':
+      if (strcmp(elt, "false") == 0) {
+        *out_data = 0;
+        continue;
+      }
+      break;
+    default:
+      break;
+    }
+
+    *lossy = true;
+    UNPROTECT(1);
+    return R_NilValue;
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP lgl_as_integer(SEXP x, bool* lossy) {
+  return Rf_coerceVector(x, INTSXP);
+}
+
+static SEXP dbl_as_integer(SEXP x, bool* lossy) {
+  double* data = REAL(x);
+  R_len_t n = Rf_length(x);
+
+  SEXP out = PROTECT(Rf_allocVector(INTSXP, n));
+  int* out_data = INTEGER(out);
+
+  for (R_len_t i = 0; i < n; ++i, ++data, ++out_data) {
+    double elt = *data;
+
+    if (elt <= INT_MIN || elt >= INT_MAX + 1.0) {
+      *lossy = true;
+      UNPROTECT(1);
+      return R_NilValue;
+    }
+
+    if (isnan(elt)) {
+      *out_data = NA_INTEGER;
+      continue;
+    }
+
+    int value = (int) elt;
+
+    if (value != elt) {
+      *lossy = true;
+      UNPROTECT(1);
+      return R_NilValue;
+    }
+
+    *out_data = value;
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP lgl_as_double(SEXP x, bool* lossy) {
+  int* data = LOGICAL(x);
+  R_len_t n = Rf_length(x);
+
+  SEXP out = PROTECT(Rf_allocVector(REALSXP, n));
+  double* out_data = REAL(out);
+
+  for (R_len_t i = 0; i < n; ++i, ++data, ++out_data) {
+    int elt = *data;
+    *out_data = elt == NA_LOGICAL ? NA_REAL : elt;
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+static SEXP int_as_double(SEXP x, bool* lossy) {
+  int* data = INTEGER(x);
+  R_len_t n = Rf_length(x);
+
+  SEXP out = PROTECT(Rf_allocVector(REALSXP, n));
+  double* out_data = REAL(out);
+
+  for (R_len_t i = 0; i < n; ++i, ++data, ++out_data) {
+    int elt = *data;
+    *out_data = elt == NA_INTEGER ? NA_REAL : elt;
+  }
+
+  UNPROTECT(1);
+  return out;
+}
+
+SEXP vec_cast(SEXP x, SEXP to) {
+  if (x == R_NilValue || to == R_NilValue) {
+    return x;
+  }
+  if (has_dim(x) || has_dim(to)) {
+    goto dispatch;
+  }
+
+  bool lossy = false;
+  SEXP out = R_NilValue;
+
+  switch (vec_typeof(to)) {
```

Comment:
Why can't you use the same double dispatch enum as in `vec_type2`?

Reply:
Because the dispatch is not symmetric for casting.

Reply:
Oooh got it.